### PR TITLE
LibGUI/PixelPaint: Introduce a new widget "DynamicWidgetContainer"

### PR DESCRIPTION
**Pull request by Torstennator**
This PR adds a new widget "DynamicWidgetContainer" that allows us to group widgets within a container widget that can be collapsed or detached to a new window. Furthermore if this widget is used as a container for other "DynamicWidgetContainers" it is possible to rearrange the possitions of the widgets within this container. All the modifications that can be done with this widget can  be persisted and reloaded automatically.

This new widget as been introduced to PixelPaint as a playground for the new feature but is easily reusable for other applications.

Details see individual commits.

Preview:

Peek 2023-11-07 19-03

### CODE REVIEW

**Comment by circl-lastname** [comment]
Personally the design doesn't fit quite right into the wider Serenity style, I'll leave that up to the maintainers though :^)

**Comment by diegogangl** [comment]
Great work @Torstennator :)
IMO this kind of widget is super important, not just for Pixel Paint but for all content editing app (and even Hack Studio). You can see it in Blender, Krita or Kdenlive; there's always going to be an overwhelming amount of tools, options and panels that need to be arranged and managed.

Some notes:
- I'd try to see how it looks without the grey backgound, just the buttons in black.
- Since you are already reserving a row for the buttons, and showing the title when they are minimized. Why not have the title of the child widget next to the buttons all the time? That way it can be removed from the child widget and save some space. Even the outline could be removed and have a single line separating the containers.

**Comment by cubiclove** [comment]
Would be awesome if there was a grab handle to rearrange the panels!

**Comment by Torstennator** [comment]
Changes since last push:
- buttons have a more decent look now
- it is now possible to rearrange sections and persist that changes (DynamicWidgetContainers within a DynmaicWidgetContainer)

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:294]: A more general thing that I was asking myself only after a while of looking at this code: Is there a reason why you're reinventing buttons and labeling and various GUI functionality and layouting calculations for the container controls? It would be much better if you could add a widget that contains the label and the control buttons. That would also allow us to change the primary direction of the container to horizontal instead of vertical, since that only requires to change the widget's layout type to vertical instead of horizontal. I think you can also apply the knurling better in this way.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:317]: This seems unnecessary and wrong.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:352]: Why does a container with individual order not have the double-click functionality? That doesn't make sense from a user perspective.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:382]: Pretty sure you can call invalidate_layout directly.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:371]: Yeah no. Please use a proper child widget for all of these controls.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:442]: Instead pass the source widget as a NNRP; the caller must check that it's passing a valid widget.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:449]: No need for abbreviation

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:461]: I'm pretty sure changes caused by modifying the child list directly are not handled correctly. Can you use the normal APIs instead?

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:470]: This empty line looks weird

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.h
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.h:24]: `MoveTargetOperation`

File: Userland/Applications/PixelPaint/PixelPaintWindow.gml
Comment (kleinesfilmroellchen) [inline Userland/Applications/PixelPaint/PixelPaintWindow.gml:52]: I don't think this is necessary if we carry over the in-window widget size correctly, as pointed out in an earlier comment.

File: Userland/Applications/PixelPaint/main.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/PixelPaint/main.cpp:58]: This should not be necessary. Please figure out how to call that function at application exit; or better yet, for each separate container widget when it's destructed (RAII).

File: Userland/Applications/PixelPaint/HistogramWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/PixelPaint/HistogramWidget.cpp:1]: The last commit can be its own pull request, since it's a separate optimization. Since I think the rest of this PR will require a lot more work, this would help with getting the fix in sooner.

File: Userland/Applications/PixelPaint/HistogramWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/PixelPaint/HistogramWidget.cpp:75]: I can't grok that lambda's name, please find something better.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
While this feature is very cool, it needs a lot more work for it to be Serenity-ready. My main points are:
- Use more non-nullable containers; please also check things that I didn't specifically point out. This reduces footguns and clarifies intent to the reader.
- Use the power of LibGUI for the container controls instead of rewriting your own primitive button controls. There's many problems with the current controls (magic numbers, bad icons, no interactivity, no keyboard motion or keyboard shortcuts possible, bad event handling, no possibility for a horizontal container mode, no respect for the knurling on/off setting) and all of those would solve themselves instantly if you just used a separate subwidget. I'd encourage you to write it in GML for simplicity; it will of course require another separate widget but the result will be much much better.
- Do not magic number any dimensions, paddings, margins, ... unless you need to account for off-by-one in graphics. We have a huge amount of magic number dimensions in the system for historical reasons, but it's nothing we should do in new code.

**Comment by Torstennator** [comment]
@kleinesfilmroellchen Thanks for the detailed review!
I'll go through all your points as i find the time to do so, just started with the first steps...

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:296]: There was a reason, initially i was aiming to have the buttons drawn as a overlay over the containing widgets and therefore i started to draw them manually as a "on_second_paint" thing. But that did not work out as expected in regards to the mouse events getting catched or propagted to the childs. With that idea the DynamicWidgetContainer would not consume space for itself.
With the current solution the "header" is taking up some space... i'll refactor that to make use of existing button and label widgets.

File: Userland/Applications/PixelPaint/HistogramWidget.cpp
Comment (Torstennator) [reply Userland/Applications/PixelPaint/HistogramWidget.cpp:1]: As this is currently not a problem, because the widget is not changing size, i would keep it with this pr as it becomes a problem with the new dynamic behavior of this pr.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:187]: That was a mistake, as i just recognized by now that the color i was looking for is the rubber band...

File: Userland/Applications/PixelPaint/PixelPaintWindow.gml
Comment (Torstennator) [reply Userland/Applications/PixelPaint/PixelPaintWindow.gml:52]: I've changed the logic so that the new window defaults to the in-window size, but having an alternative other detached size seems quite comfortable as you might want to detach the widget because it should display more information when more space is available - so you don't need to resize the window manually after detaching...

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:112]: Currently my logic is to persist the view state or the order of child widgets whereas to order is also some kind of view state. So i use one or the other based on `is_container_with_individual_order()`.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: Since there is now a new widget that has all the buttons and the label - this is now obsolete.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: For the moment i removed the knurl stuff - not sure if i liked it.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:222]: Was thinking about it, where is already solved that on pixel paint.... that kind of widgets would need to be wrapped in another widget, so that the visibility does not get screwed up by the container.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: A new widget that contains the label and buttons has been implemented.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: If that is not in place, the mouse cursor will only change back if you move the mouse cursor.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: Iv'e enabled the double click again, there was no real reason for that - i just didn't liked the collapsed main section in pixelpaint - but that should not be a general restriction.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:371]: Hm, it is not working as expected without the deferred invoke and resulted in the widgets not always being drawn completely. So I didn't changed the existing solution as that working as expected.

File: Userland/Applications/PixelPaint/main.cpp
Comment (Torstennator) [reply Userland/Applications/PixelPaint/main.cpp:58]: For the moment i was not able to find a reliable working solution e.g. closing all windows within the destructor. That destructor somehow never was called when the widget has been cunstructed via a gml file - I have no idea whats the reason behind that.
For the moment i close all the open windows in the destructor as when the widget was not constructed via gml and the old solution where they are closed via the static function.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: For the moment i really want the have the window independent from the main window and not blocking the main window while it's active.
A modal window that e.g. blocks the main window could probably be some future extension if that is needed.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: I was trying to change that, but the compiler was not happy about the const so i've not changed that as suggested.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: I was trying to improve that by using `remove_child` and `insert_child_before` but that didn't worked out. Somehow it was never working as expected and introduced some weird looking spaces between the moved widgets. For to moment i would like to leave the not perfect but working solution :-/

**Comment by kleinesfilmroellchen** [comment]
Excellent progress! On my phone so a full review is hard, but a few points to consider:
- Optional<NNRP> can just be RP and Optional<RP> also is just a RP.
- Use static GML generation, dynamic GML generation is for Playground only.
- The labels and buttons shouldn't need a fixed height since they can correctly calculate their min height. (They are already "shrink")
- Check spelling and grammar in the new docs.

**Comment by Torstennator** [comment]
Hey @kleinesfilmroellchen ,
I've addressed your latest remarks and hope to get the pr closer to a state where i can remove the "Draft" state. Let me know if you have any further concerns.

**Comment by circl-lastname** [comment]
Bump, this is too cool

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:237]: Mixing and matching `m_detached_size` and `detached_size()` feels awkward. Maybe `detached_size()` should just return an Optional size always?

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:261]: Avoid `this->` when it's not needed.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:246]: `auto& widget: childs` should be fine and avoid copying the refptrs.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:247]: `this->` isn't necessary.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:315]: "curse" -> "cursor"

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:343]: `DynamicWidgetContainerControls::get_event_dispatcher()`? Though, maybe that method should have a better name. (`get_section_label()`?)

It also feels a little weird to be reaching inside and modifying its child widgets instead of telling it to do so. But no big deal.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:212]: What's the reason for making this a gradient? It seems a little arbitrary.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:218]: The reason you can't is that the `painter(*this)` constructor above is clipping to this widget's bounds. I think `painter.clear_clip_rect()` would solve this but I haven't tested it myself. Might take some fiddling with.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.h
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DynamicWidgetContainer.h:45]: :thinking: Is this true for all widgets created from GML? If you can find a reduced test case it would be useful to create an issue about this. (Even if you can't reduce it, create an issue with whatever you know.)

File: Userland/Applications/PixelPaint/PixelPaintWindow.gml
Comment (cubiclove) [inline Userland/Applications/PixelPaint/PixelPaintWindow.gml:47]: Would it be better to rename this to 'Editor Panels', 'Studio Panels' or something?
Currently the box on the left with all of the brushes and selection tools is called the Tool Panel, so it might be a little confusing (in terms of documentation).

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: As the compiler was not happy with the name `child_containers` due to interference with the function name an auto deduction i changed the name to `containers`.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: That is a bit clunky because it is used as a size property and therfore does not support the optional :/ I've introduced a `has_detached_size()` function.

File: Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp
Comment (Torstennator) [reply Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp]: There is no reason, that was just how i would like to draw a small separation between the widgets . I've changed that to a solid line.

**Comment by Torstennator** [comment]
> This is very, very cool and something we really need! :^)
> 
> Most of this feedback is superficial things, generally this seems good to me.
> 
> I have noticed oddness when playing around with this though. When launching Pixel Paint, some tool windows will appear slightly too small and without their widgets being visible. Collapsing/expanding them, or detaching them, doesn't help. For example, I have this with the "Tool Properties" widget. I have to switch tool to make its contents appear. It's the same if I show the histogram widget, it's too small and doesn't display the histogram.
> 
> Also, there's mention of knurling in the code, but I don't see any? Knurling is specifically the visual stripes to show something is resizable or moveable.

I've double checked the issue with the "Tool Properties" and the same behaviour is currently on the master branch. It only shows the properties if you change the tool at least once - so that seems to be nothing new but should be fixed to display the properties for the initial selected tool without the need to change the tool.
Regarding the histogram widget - i was not able to reproduce that - but with this pr i also changed the logic for the histogram widget so that it only shows something if the image has some data to display.
The knureling was used while developing this widget but i got rid of rendering them while making review changes - now the m_move_widget_knurl is only used as the region where the cursor changes to a "Hand" to indicate that it can be moved.

**Comment by Torstennator** [comment]
Hi @AtkinsSJ, I'm not sure if there is still a todo on my side. I don't see any open topics for this pr but the status tells me that there is still a requested change from you.
Please let me know if there is still some open stuff...
Greetings

**Pull request merged**
